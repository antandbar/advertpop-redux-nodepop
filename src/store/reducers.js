import {
    AUTH_LOGIN_FAILURE,
    AUTH_LOGIN_REQUEST,
    AUTH_LOGIN_SUCCESS,
    AUTH_LOGOUT_SUCCESS,
    ADVERTS_LOADED_REQUEST,
    ADVERTS_LOADED_SUCCESS,
    ADVERTS_LOADED_FAILURE,
    UI_RESET_ERROR,
    ADVERT_LOADED_SUCCESS,
    ADVERT_LOADED_REQUEST,
    ADVERT_LOADED_FAILURE,
    ADVERT_CREATED_REQUEST,
    ADVERT_CREATED_SUCCESS,
    ADVERT_CREATED_FAILURE,
    TAGS_LOADED_REQUEST,
    TAGS_LOADED_SUCCESS,
    TAGS_LOADED_FAILURE,
    ADVERT_DELETED_REQUEST,
    ADVERT_DELETED_SUCCESS,
    ADVERT_DELETED_FAILURE
  } from './types';
  
  export const defaultState = {
    auth: false,
    adverts: {
      loaded: false,
      data: [],
    },
    ui: {
      isLoading: false,
      error: null,
    },
    tags: {
      loaded: false,
      data: [],
    }
  };
  
  export const auth = (state = defaultState.auth, action) => {
    switch (action.type) {
      case AUTH_LOGIN_SUCCESS:
        return true;
      case AUTH_LOGOUT_SUCCESS:
        return false;
      default:
        return state;
    }
  };
  
  export const adverts = (state = defaultState.adverts, action) => {
    switch (action.type) {
      case ADVERTS_LOADED_SUCCESS:
        return { loaded: true, data: action.payload };
      case ADVERT_LOADED_SUCCESS:
        return { ...state, data: [...state.data, action.payload] };
        case ADVERT_CREATED_SUCCESS:
          return { ...state, data: [...state.data, action.payload] };
      case ADVERT_DELETED_SUCCESS: 
        return {...state, data: state.data.filter(advert => advert.id !== action.payload)}; 
      default:
        return state;
    }
  };

  export const tags = (state = defaultState.tags, action) => {
    switch (action.type) {
      case TAGS_LOADED_SUCCESS:
        return { loaded: true, data: action.payload };
      default:
        return state;
    }
  };
  
  export const ui = (state = defaultState.ui, action) => {
    switch (action.type) {
      case AUTH_LOGIN_REQUEST:
      case ADVERTS_LOADED_REQUEST:
      case TAGS_LOADED_REQUEST:
      case ADVERT_LOADED_REQUEST:
      case ADVERT_CREATED_REQUEST:
      case ADVERT_DELETED_REQUEST:
        return { ...state, isLoading: true, error: null };
      case AUTH_LOGIN_SUCCESS:
      case ADVERTS_LOADED_SUCCESS:
      case TAGS_LOADED_SUCCESS:
      case ADVERT_LOADED_SUCCESS:
      case ADVERT_CREATED_SUCCESS:
      case ADVERT_DELETED_SUCCESS:
        return { ...state, isLoading: false };
      case AUTH_LOGIN_FAILURE:
      case ADVERTS_LOADED_FAILURE:
      case TAGS_LOADED_FAILURE:
      case ADVERT_LOADED_FAILURE:
      case ADVERT_CREATED_FAILURE:
      case ADVERT_DELETED_FAILURE:
        return { ...state, isLoading: false, error: action.payload };
      case UI_RESET_ERROR:
        return { ...state, error: null };
      default:
        return state;
    }
  };