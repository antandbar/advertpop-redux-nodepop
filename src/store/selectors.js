export const getIsLogged = state => state.auth;

export const getAdverts = state => state.adverts.data;

export const getAdvertsFilter =
  (nameFilter, saleFilter, rangeFilter, multiSelectorFilter) => state => {
    const result = state.adverts.data.filter(
      advert =>
        (nameFilter.length ? advert.name === nameFilter : true) &&
        (saleFilter.length ? advert.sale.toString() === saleFilter : true) &&
        advert.price < Number(rangeFilter.rangeMax) &&
        advert.price > Number(rangeFilter.rangeMin) &&
        (multiSelectorFilter.length
          ? advert.tags.every(item => multiSelectorFilter.includes(item))
          : true),
    );

    return result;
  };

export const getAreAdvertsLoaded = state => state.adverts.loaded;

export const getTags = state => state.tags.data;

export const getAreTagsLoaded = state => state.tags.loaded;

export const getAdvert = advertId => state =>
  state.adverts.data.find(advert => advert.id === advertId);

export const getUi = state => state.ui;
