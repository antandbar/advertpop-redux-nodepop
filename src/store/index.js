import { createStore, combineReducers, applyMiddleware } from 'redux';
import * as reducers from './reducers';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';
import * as auth from '../components/auth/service';
import * as adverts from '../components/adverts/service';
import { AUTH_LOGIN_SUCCESS, ADVERT_CREATED_SUCCESS } from './types';

const api = { auth, adverts };

const timestamp = () => next => action => {
  const newAction = {
    ...action,
    meta: {
      ...action.meta,
      timestamp: new Date(),
    },
  };
  return next(newAction);
};

const successRedirections =
  (history, redirections) => _store => next => action => {
    const result = next(action);
    const redirection = redirections[action.type];
    if (redirection) {
      redirection(history, action.payload);
    }

    return result;
  };

const configureStore = (preloadedState, { history }) => {
  const middlewares = [
    thunk.withExtraArgument({ api, history }),
    successRedirections(history, {
      [AUTH_LOGIN_SUCCESS]: history => {
        const from = history.location.state?.from?.pathname || '/';
        history.replace(from);
      },
      [ADVERT_CREATED_SUCCESS]: (history, payload) =>
        history.push(`/adverts/${payload.id}`), 
    }),
    timestamp,
  ];

  const store = createStore(
    combineReducers(reducers),
    preloadedState,
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  return store;
};

export default configureStore;
