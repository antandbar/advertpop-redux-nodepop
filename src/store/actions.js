import { getAreAdvertsLoaded, getAdvert, getAreTagsLoaded } from './selectors';
import {
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCESS,
  AUTH_LOGOUT_SUCCESS,
  ADVERTS_LOADED_FAILURE,
  ADVERTS_LOADED_REQUEST,
  ADVERTS_LOADED_SUCCESS,
  ADVERT_CREATED_SUCCESS,
  ADVERT_CREATED_FAILURE,
  ADVERT_CREATED_REQUEST,
  ADVERT_LOADED_SUCCESS,
  ADVERT_LOADED_FAILURE,
  ADVERT_LOADED_REQUEST,
  UI_RESET_ERROR,
  TAGS_LOADED_FAILURE,
  TAGS_LOADED_REQUEST,
  TAGS_LOADED_SUCCESS,
  ADVERT_DELETED_SUCCESS,
  ADVERT_DELETED_FAILURE,
  ADVERT_DELETED_REQUEST
} from './types';

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginSuccess = () => ({
  type: AUTH_LOGIN_SUCCESS,
});

export const authLoginFailure = error => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});

export const authLogin = credentials => {
  return async function (dispatch, _getState, { api, history }) {
    dispatch(authLoginRequest());
    try {
      await api.auth.login(credentials);
      dispatch(authLoginSuccess());         
    } catch (error) {
      dispatch(authLoginFailure(error));
      throw error;
    }
  };
};

export const authLogoutSuccess = () => ({
  type: AUTH_LOGOUT_SUCCESS,
});

export const authLogout = () => {
  return function (dispatch, _getState, { api }) {
    return api.auth.logout().then(() => {
      dispatch(authLogoutSuccess());
    });
  };
};

export const advertsLoadedRequest = () => ({
  type: ADVERTS_LOADED_REQUEST,
});

export const advertsLoadedSuccess = adverts => ({
  type: ADVERTS_LOADED_SUCCESS,
  payload: adverts,
});

export const advertsLoadedFailure = error => ({
  type: ADVERTS_LOADED_FAILURE,
  payload: error,
  error: true,
});

export const advertsLoaded = () => {
  return async function (dispatch, getState, { api }) {
    const advertsLoaded = getAreAdvertsLoaded(getState());
    if (advertsLoaded) return;

    dispatch(advertsLoadedRequest());
    try {
      const adverts = await api.adverts.getAdverts();
      dispatch(advertsLoadedSuccess(adverts));
    } catch (error) {
      dispatch(advertsLoadedFailure(error));
      throw error;
    }
  };
};

export const tagsLoadedRequest = () => ({
  type: TAGS_LOADED_REQUEST,
});

export const tagsLoadedSuccess = adverts => ({
  type: TAGS_LOADED_SUCCESS,
  payload: adverts,
});

export const tagsLoadedFailure = error => ({
  type: TAGS_LOADED_FAILURE,
  payload: error,
  error: true,
});

export const tagsLoaded = () => {
  return async function (dispatch, getState, { api }) {
    const tagsLoaded = getAreTagsLoaded(getState());
    if (tagsLoaded) return;

    dispatch(tagsLoadedRequest());
    try {
      const tags = await api.adverts.getTags();
      dispatch(tagsLoadedSuccess(tags));
    } catch (error) {
      dispatch(tagsLoadedFailure(error));
      throw error;
    }
  };
};

export const advertLoadedRequest = () => ({
  type: ADVERT_LOADED_REQUEST,
});

export const advertLoadedSuccess = advert => ({
  type: ADVERT_LOADED_SUCCESS,
  payload: advert,
});

export const advertLoadedFailure = error => ({
  type: ADVERT_LOADED_FAILURE,
  payload: error,
  error: true,
});

export const advertLoaded = (advertId) => {
  return async function (dispatch, getState, { api, history }) {
    const advertLoaded = getAdvert(advertId)(getState());
    if (advertLoaded) return;
    dispatch(advertLoadedRequest());
    try {
      const advert = await api.adverts.getAdvert(advertId);
      dispatch(advertLoadedSuccess(advert));
    } catch (error) {
      dispatch(advertLoadedFailure(error));
      history.push('/404');
    }
  };
};

export const advertCreatedRequest = () => ({
  type: ADVERT_CREATED_REQUEST,
});

export const advertCreatedSuccess = (advert) => ({
  type: ADVERT_CREATED_SUCCESS,
  payload: advert
});

export const advertCreatedFailure = error => ({
  type: ADVERT_CREATED_FAILURE,
  payload: error,
  error: true,
});

export const advertCreated = (advert) => {
  return async function (dispatch, getState, { api }) {
    dispatch(advertCreatedRequest());
    try {
      const createdAdvert = await api.adverts.createAdvert(advert);
      dispatch(advertCreatedSuccess(createdAdvert));
    } catch (error) {
      dispatch(advertCreatedFailure(error));
      throw error;
    }
  };
};

export const advertDeletedRequest = () => ({
  type: ADVERT_DELETED_REQUEST,
});

export const advertDeletedSuccess = advertId => ({
  type: ADVERT_DELETED_SUCCESS,
  payload: advertId,
});

export const advertDeletedFailure = error => ({
  type: ADVERT_DELETED_FAILURE,
  payload: error,
  error: true,
});

export const advertDeleted = advertId => {
  return async function (dispatch, getState, {  api, history  }) {
    dispatch(advertDeletedRequest());
    try {
      await api.adverts.deleteAdvert(advertId);
      dispatch(advertDeletedSuccess(advertId));
      history.push('/adverts');
    } catch (error) {
      dispatch(advertDeletedFailure(error));
      throw error;
    }
  };
};

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
