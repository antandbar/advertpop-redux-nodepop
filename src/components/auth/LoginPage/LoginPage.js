import { useState, useRef, useEffect } from 'react';
import Button from '../../common/Button';
import FormField from '../../common/FormField';
import './LoginPage.css';
import { useDispatch, useSelector } from 'react-redux';
import { authLogin, uiResetError } from '../../../store/actions';
import { getUi } from '../../../store/selectors';

function LoginPage() {
  const ref = useRef(null);
  const [credentials, setCredentials] = useState({
    email: '',
    password: '',
    remember: false,
  });

  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  useEffect(() => {
    ref.current.focus();
  }, []);

  const { email, password, remember } = credentials;

  const handleChange = ({ target: { value, name, type, checked } }) => {
    setCredentials(credentials => ({
      ...credentials,
      // si es Chekbox se toma checked, si no es así value
      [name]: type === 'checkbox' ? checked : value,
    }));
  };

  const resetError = () => dispatch(uiResetError());

  const handleSubmit = event => {
    event.preventDefault();
    dispatch(authLogin(credentials))
  };

  return (
    <div className="loginPage">
      <h1 className="loginPage-title">Log in to AdvertPop</h1>
      <form onSubmit={handleSubmit}>
        <FormField
          type="email"
          name="email"
          label="email"
          className="loginForm-field"
          value={email}
          onChange={handleChange}
          ref={ref}
        />
        <FormField
          type="password"
          name="password"
          label="password"
          className="loginForm-field"
          value={password}
          onChange={handleChange}
        />

        <input
          type="checkbox"
          name="remember"
          checked={remember}
          value="remember"
          onChange={handleChange}
        />
        <label>Recordar contraseña</label>

        <Button
          className="loginForm-submit"
          type="submit"
          variant="primary"
          disabled={!email || !password || isLoading}
        >
          Log in
        </Button>
      </form>
      {error && (
        <div onClick={resetError} className="loginPage-error">
          {error.message}
        </div>
      )}
    </div>
  );
}

export default LoginPage;
