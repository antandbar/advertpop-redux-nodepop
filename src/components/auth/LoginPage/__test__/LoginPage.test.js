import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import LoginPage from '../LoginPage';
import { authLogin, uiResetError } from '../../../../store/actions';



jest.mock('../../../../store/actions');

describe('LoginPage', () => {
  test('snapshot', () => {
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    };
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    const { container } = render(
      <Provider store={store}>
          <LoginPage />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('should authLogin action', () => {
    authLogin.mockReturnValue('action');
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    };
    const email = 'antonio';
    const password = '123456';
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
  
    render(
      <Provider store={store}>
        <LoginPage/>
      </Provider>,
    );
    const emailInput = screen.getByLabelText(/email/);
    const passwordInput = screen.getByLabelText(/password/);
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();
    fireEvent.change(emailInput, { target: { value: email } });
    fireEvent.change(passwordInput, { target: { value: password } });
    expect(submitButton).not.toBeDisabled();



    fireEvent.click(submitButton);
    const credentials = authLogin.mock.calls[0][0];
    expect(credentials).toMatchObject({ email, password });
    expect(store.dispatch).toHaveBeenCalledWith('action');
  });

  test('should reset error', () => {
    const error = { message: 'Network error' };
    const state = {
      ui: {
        error,
        isLoading: false,
      },
    };
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    render(
      <Provider store={store}>
          <LoginPage />
      </Provider>,
    );
  
    const errorElement = screen.getByText(error.message);
    expect(errorElement).toBeInTheDocument();
    expect(screen.queryByText('not exist')).not.toBeInTheDocument();
    fireEvent.click(errorElement);
    expect(uiResetError).toHaveBeenCalled();
  });
});
