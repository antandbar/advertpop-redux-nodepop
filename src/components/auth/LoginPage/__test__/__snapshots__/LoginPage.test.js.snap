// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div
    class="loginPage"
  >
    <h1
      class="loginPage-title"
    >
      Log in to AdvertPop
    </h1>
    <form>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            email
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="email"
            type="email"
            value=""
          />
        </label>
      </div>
      <div
        class="formField loginForm-field"
      >
        <label
          class="formField-label"
        >
          <span>
            password
          </span>
          <input
            autocomplete="off"
            class="formField-input"
            name="password"
            type="password"
            value=""
          />
        </label>
      </div>
      <input
        name="remember"
        type="checkbox"
        value="remember"
      />
      <label>
        Recordar contrase√±a
      </label>
      <button
        class="sc-bdvvtL doPQTC loginForm-submit"
        disabled=""
        type="submit"
      >
        Log in
      </button>
    </form>
  </div>
</div>
`;
